### PR TITLE
add api response examples to the api documentation

### PR DESCRIPTION
**Pull request by MaximilianMeister**
screenshot from 2016-09-27 08-43-34

### CODE REVIEW

**Comment by MaximilianMeister** [comment]
[(5) #upgrade Synchronize API responses with the Angular app](https://trello.com/c/3TALQwDA/121-5-upgrade-synchronize-api-responses-with-the-angular-app)

File: crowbar_framework/app/controllers/repositories_controller.rb [hunk 1]
```diff
@@ -135,11 +158,7 @@ def activate_all
     end
   end
 
-  #
-  # Deactivate all Repositories
-  #
-  # Provides the restful api call for
-  # Deactivate all Repositories   /utils/repositories/deactivate_all   POST   Destroys Repository DataBagItem
+  api :POST, "/utils/repositories/deactivate_all", "Dectivate all repositories. Destroys DataBagItems"
```
Comment (MaximilianMeister) [inline crowbar_framework/app/controllers/repositories_controller.rb]: fixed

File: crowbar_framework/app/controllers/repositories_controller.rb [hunk 2]
```diff
@@ -93,12 +122,10 @@ def activate
     end
   end
 
-  #
-  # Deactivate a single Repository
-  #
-  # Provides the restful api call for
-  # Deactivate a Repository   /utils/repositories/deactivate   POST   Destroys Repository DataBagItem
-  # required parameters: platform, arch, repo
+  api :POST, "/utils/repositories/deactivate", "Deactivate a single repository. Destroys a DataBagItem"
```
Comment (MaximilianMeister) [inline crowbar_framework/app/controllers/repositories_controller.rb]: fixed

File: crowbar_framework/app/controllers/api/crowbar_controller.rb [hunk 3]
```diff
@@ -54,6 +77,24 @@ def maintenance
 
   api :GET, "/api/crowbar/repocheck", "Sanity check for Crowbar server repositories"
   api_version "2.0"
+  example '
```
Comment (jsuchome) [inline crowbar_framework/app/controllers/api/crowbar_controller.rb]: Have we already agreed on the format of the result here?

Comment (MaximilianMeister) [inline crowbar_framework/app/controllers/api/crowbar_controller.rb]: no, but do we need to show something else besides the json response?

File: crowbar_framework/app/controllers/barclamp_controller.rb [hunk 4]
```diff
@@ -255,6 +437,63 @@ def proposals
   api :GET, "/crowbar/:barclamp/1.0/proposals/template",
     "Returns the content of a proposal template"
   param :barclamp, String, desc: "Name of the barclamp", required: true
+  example '
+  {
+    "id": "template-dns",
```
Comment (santiph) [inline crowbar_framework/app/controllers/barclamp_controller.rb]: A proposal template for each type of barclamp supported must be documented

**Comment by santiph** [comment]
Apipie Headers example: http://stackoverflow.com/questions/27028393/how-to-generate-header-for-apipie-rails-gem
https://github.com/Apipie/apipie-rails/issues/340
https://github.com/Apipie/apipie-rails/pull/341

File: crowbar_framework/app/controllers/api/backups_controller.rb [hunk 5]
```diff
@@ -20,12 +20,36 @@ class Api::BackupsController < ApiController
...
+  ]
+  '
   def index
     render json: Api::Backup.all
   end
 
   api :GET, "/api/crowbar/backups/:id", "Returns a specific backup"
   api_version "2.0"
+  example '
+  {
```
Comment (MaximilianMeister) [inline crowbar_framework/app/controllers/api/backups_controller.rb]: with the download endpoint

File: crowbar_framework/app/controllers/api/backups_controller.rb [hunk 6]
```diff
@@ -35,6 +59,17 @@ def show
...
+    "id": 1,
+    "name": "testbackup",
+    "version": "4.0",
+    "size": 76815,
+    "created_at": "2016-09-27T06:05:10.208Z",
+    "updated_at": "2016-09-27T06:05:10.208Z",
+    "migration_level": 20160819142156
+  }
```
Comment (MaximilianMeister) [inline crowbar_framework/app/controllers/api/backups_controller.rb]: why that? you mean it should be downloaded right after creation? I think that would be wrong

File: crowbar_framework/app/controllers/api/crowbar_controller.rb [hunk 7]
```diff
@@ -30,6 +39,20 @@ def update
...
   api :GET, "/api/crowbar/upgrade", "Status of Crowbar Upgrade"
+  example '
+  {
+    "version": "4.0",
+    "addons": [
+      "ceph",
+      "ha"
+    ],
+    "upgrade": {
+      "upgrading": false,
```
Comment (MaximilianMeister) [inline crowbar_framework/app/controllers/api/crowbar_controller.rb]: ok, I can add a comment next to it

File: crowbar_framework/app/controllers/api/upgrade_controller.rb [hunk 8]
```diff
@@ -19,6 +19,29 @@ class Api::UpgradeController < ApiController
 
   api :GET, "/api/upgrade", "Show the Upgrade status object"
```
Comment (MaximilianMeister) [inline crowbar_framework/app/controllers/api/upgrade_controller.rb]: just look at the example :) `/api/crowbar/upgrade` doesn't include the checks

File: crowbar_framework/app/controllers/barclamp_controller.rb [hunk 9]
```diff
@@ -40,6 +40,39 @@ def controller_to_barclamp
 
   add_help(:barclamp_index)
   api :GET, "/crowbar", "Returns a list of string names and descriptions for all barclamps"
+  example '
```
Comment (MaximilianMeister) [inline crowbar_framework/app/controllers/barclamp_controller.rb]: no. why do you think that?

File: crowbar_framework/app/controllers/barclamp_controller.rb (hunk 4 above)
Comment (MaximilianMeister) [inline crowbar_framework/app/controllers/barclamp_controller.rb]: that would be too much IMO, as this is just an example. does it really need to be that detailed? It would significantly bloat the LOC of the barclamp_controller

File: crowbar_framework/app/controllers/api/errors_controller.rb [hunk 10]
```diff
@@ -20,13 +20,39 @@ class Api::ErrorsController < ApiController
 
   api :GET, "/api/errors", "Show a list of errors"
   api_version "2.0"
+  example '
+  [
+    {
+      "id": 1,
+      "error": "StandardError",
+      "message": "This is an error",
+      "code": null,
```
Comment (MaximilianMeister) [inline crowbar_framework/app/controllers/api/errors_controller.rb]: > Apipie Headers example: http://stackoverflow.com/questions/27028393/how-to-generate-header-for-apipie-rails-gem
> Apipie/apipie-rails#340
> Apipie/apipie-rails#341

that would be done in another PR

**Comment by MaximilianMeister** [comment]
> Apipie Headers example: http://stackoverflow.com/questions/27028393/how-to-generate-header-for-apipie-rails-gem
Apipie/apipie-rails#340
Apipie/apipie-rails#341

that would be done in another PR

File: crowbar_framework/app/controllers/api/errors_controller.rb (hunk 10 above)
Comment (MaximilianMeister) [inline crowbar_framework/app/controllers/api/errors_controller.rb]: added an example where you can see the datatype

File: crowbar_framework/app/controllers/api/crowbar_controller.rb [hunk 11]
```diff
@@ -19,6 +19,15 @@ class Api::CrowbarController < ApiController
 
   api :GET, "/api/crowbar", "Show the crowbar object"
   api_version "2.0"
+  example '
+  {
+    "version": "4.0",
```
Comment (MaximilianMeister) [inline crowbar_framework/app/controllers/api/crowbar_controller.rb]: cloud 6 is crowbar 3.0, cloud 7 is crowbar 4.0

Comment (santiph) [inline crowbar_framework/app/controllers/api/crowbar_controller.rb]: Please Add that as part of the documentation

File: crowbar_framework/app/controllers/barclamp_controller.rb (hunk 9 above)
Comment (santiph) [inline crowbar_framework/app/controllers/barclamp_controller.rb]: I'd assume barclamp is an attribute of the crowbar entity, different than Version, for instance

File: crowbar_framework/app/controllers/barclamp_controller.rb (hunk 4 above)
Comment (santiph) [inline crowbar_framework/app/controllers/barclamp_controller.rb]: You cannot provide information through the API that is not being documented.
In other words, you cannot expect attributes or fields are being consumed by a client if they're not documented somewhere.

Comment (MaximilianMeister) [inline crowbar_framework/app/controllers/barclamp_controller.rb]: in an ideal world I agree, that'd be super useful. but this would be an enormous amount of documentation, just imagine a proposal alone, and then repository metadata etc. we don't have the time to do that now

File: crowbar_framework/app/controllers/api/crowbar_controller.rb (hunk 11 above)
Comment (MaximilianMeister) [inline crowbar_framework/app/controllers/api/crowbar_controller.rb]: where? as a comment on that line it would be probably too long

File: crowbar_framework/app/controllers/barclamp_controller.rb (hunk 9 above)
Comment (MaximilianMeister) [inline crowbar_framework/app/controllers/barclamp_controller.rb]: this is the v1 api, I think you are referring to v2 `/api/crowbar`

File: crowbar_framework/app/controllers/barclamp_controller.rb (hunk 4 above)
Comment (santiph) [inline crowbar_framework/app/controllers/barclamp_controller.rb]: You can create tech debts to track that work separately and even in a different sprint, as long as PO (Vincent) is ok with it. I'm only raising issues as I see them.

File: crowbar_framework/app/controllers/api/upgrade_controller.rb (hunk 8 above)
Comment (santiph) [inline crowbar_framework/app/controllers/api/upgrade_controller.rb]: So `/api/upgrade` also contains `/api/crowbar/upgrade` results? Could you add that as a comment in the example?

File: crowbar_framework/app/controllers/barclamp_controller.rb (hunk 9 above)
Comment (santiph) [inline crowbar_framework/app/controllers/barclamp_controller.rb]: Good point. Thanks

File: crowbar_framework/app/controllers/api/crowbar_controller.rb (hunk 11 above)
Comment (santiph) [inline crowbar_framework/app/controllers/api/crowbar_controller.rb]: I'd prefer a long line in the example than a comment on a github review ;)
You can also add those lines outside the json example and make it multiline... 

I'm reading Apipie issues and it seems there is no better approach to represent the responses other than examples. A lot of people reference to swagger.io or raml.org, claiming they both support it...

File: crowbar_framework/app/controllers/api/backups_controller.rb (hunk 6 above)
Comment (santiph) [inline crowbar_framework/app/controllers/api/backups_controller.rb]: Well, I guess we can use the id to download the backup file. Although I would expect an href > download > link attribute to get it from this response

File: crowbar_framework/app/controllers/barclamp_controller.rb (hunk 4 above)
Comment (santiph) [inline crowbar_framework/app/controllers/barclamp_controller.rb]: well... This is not affecting our work right now, so I can still approve it if a tech debt to track this work is created and linked

File: crowbar_framework/app/controllers/api/upgrade_controller.rb (hunk 8 above)
Comment (MaximilianMeister) [inline crowbar_framework/app/controllers/api/upgrade_controller.rb]: just ignore it for now, `/api/upgrade` will change this sprint anyways, as soon as we start working on https://trello.com/c/OWehKX6K/126-5-implement-a-state-machine-for-the-upgrade

File: crowbar_framework/app/controllers/barclamp_controller.rb (hunk 4 above)
Comment (MaximilianMeister) [inline crowbar_framework/app/controllers/barclamp_controller.rb]: as I said there will be follow ups, I really want to avoid those huge PR's. but as this is a significant amount of work which is not necessary for the upgrade process, we have to postpone it.

the card to track this is https://trello.com/c/gs8pbS7M/1170-enhance-api-documentation

File: crowbar_framework/app/controllers/repositories_controller.rb [hunk 12]
```diff
@@ -64,12 +95,10 @@ def sync
...
-  #
-  # Activate a single Repository
-  #
-  # Provides the restful api call for
-  # Activate a Repository   /utils/repositories/activate   POST  Creates Repository DataBagItem
-  # required parameters: platform, arch, repo
+  api :POST, "/utils/repositories/activate", "Activate a single repository. Creates a DataBagItem"
+  param :platform, String, desc: "Platform of the repository", required: true
+  param :arch, String, desc: "Architecture of the repository", required: true
+  param :repo, String, desc: "Name of the repository", required: true
```
Comment (MaximilianMeister) [inline crowbar_framework/app/controllers/repositories_controller.rb]: there is no response here in case of success, only 200 OK (like i think in all POST requests in crowbar, except backup create)

**Comment by MaximilianMeister** [comment]
@santiph some comments are really implementation specific, like the backup create response, or would require changes there. this PR is really just to define the status quo in terms of documentation.

for the inline comments bit, I am not sure if that doesn't clutter the response examples too much. we may add some extra info below the json response, but to say it again, lets keep things simple and incrementally open smaller PR's to make the reviewers easier

File: crowbar_framework/app/controllers/api/crowbar_controller.rb (hunk 3 above)
Comment (MaximilianMeister) [inline crowbar_framework/app/controllers/api/crowbar_controller.rb]: @jsuchome @santiph as mentioned below, we may have to define some view to show extra information about response parameters that are unclear. but please in a separate small PR

**Comment by vuntz** [comment]
@santiph let's not use this pull request as a way to change the APIs, as otherwise, this will never get merged (and tracking this here is a very good way to make sure that we will forget about some requested changes).

This pull request is just about documenting better the current behavior of the APIs, and once we have that, we can have some good discussion about what should be changed and how.

File: crowbar_framework/app/controllers/api/upgrade_controller.rb (hunk 8 above)
Comment (santiph) [inline crowbar_framework/app/controllers/api/upgrade_controller.rb]: Fair enough

File: crowbar_framework/app/controllers/repositories_controller.rb (hunk 12 above)
Comment (santiph) [inline crowbar_framework/app/controllers/repositories_controller.rb]: HTTP Codes returned needs also to be documented. Later on, you can also use http codes for specific backend errors, when described

File: crowbar_framework/app/controllers/api/crowbar_controller.rb (hunk 11 above)
Comment (aspiers) [inline crowbar_framework/app/controllers/api/crowbar_controller.rb]: I agree with @santiph. Github PR comments should never be considered as documentation ...

File: crowbar_framework/app/controllers/repositories_controller.rb (hunk 12 above)
Comment (MaximilianMeister) [inline crowbar_framework/app/controllers/repositories_controller.rb]: done now in https://github.com/crowbar/crowbar-core/pull/712

File: crowbar_framework/app/controllers/api/crowbar_controller.rb (hunk 11 above)
Comment (MaximilianMeister) [inline crowbar_framework/app/controllers/api/crowbar_controller.rb]: @aspiers me too :) not sure who would not ... :/
